dali/python/nvidia/dali/pool.py

```diff
@@ -0,0 +1,463 @@
+        return batch_i not in self.partially_received
+
+    def is_not_received(self, batch_i, tasks):
+        """Check if we didn't receive all results for given tasks and batch_i"""
+        return len(self.partially_received[batch_i]) < len(tasks)
+
+    def get_batch(self, batch_i, tasks):
+        """Return the full batch, mark it as cleared and consumed"""
+        full_batch = self.partially_received[batch_i]
+        res = [full_batch[i] for i, _ in tasks]
+        del full_batch
+        return res
+
+    def receive_chunk(self, batch_i, sock, batch_serialized):
+        """Obtain the chunk and decode it, add to partially gathered result"""
+        worker_batch = self.batch_consumer.load_batch(sock, batch_serialized)
+        self.partially_received[batch_i].update(worker_batch)
+
+
+class ProcPool:
+    """Runs pool of worker processes, stores pipes and sockets used to communicate with the workers,
+starts thread keeping track of running processes and initializes communication.
+"""
+
+    def __init__(
+            self, callbacks, prefetch_queue_depths, workers_num=None, init_method="fork",
+            initial_chunk_size=1024 * 1024):
+        if len(callbacks) != len(prefetch_queue_depths):
+            raise RuntimeError("Number of prefetch queues must match number of callbacks")
+        if any(prefetch_queue_depth <= 0 for prefetch_queue_depth in prefetch_queue_depths):
+            raise RuntimeError("Prefetch queue must have at least one element")
+        if initial_chunk_size <= 0:
+            raise RuntimeError("Chunk capacity must be positive integer")
+        if init_method == 'fork' and _b.HasCudaContext():
+            raise RuntimeError(
+                "Cannot fork process when there is CUDA context bound to it. "
+                "Make sure you build pipeline before CUDA context is acquired or change Python workers "
+                "starting method from ``fork`` to ``spawn`` (see Piepline's ``py_workers_init`` option). "
+                "If you are trying to build multiple pipelines that use Python workers, you will need to "
+                "call ``start_py_workers`` method on all of them before calling ``build`` method of any pipeline.")
+        mp = multiprocessing.get_context(init_method)
+        if worker is None:
+            workers_num = mp.cpu_count()
+        if workers_num < 1:
+            raise RuntimeError("workers_num must be a positive integer")
+        self._workers_num = workers_num
+        self._processes = []
+        self._task_pipes = []
+        self._res_pipes = []
+        self._socks = []
+        self._task_pipes_lock = threading.Lock()
+        self._from_tracker = None
+        self._to_tracker = None
+        self._tracker_thread = None
+        for i in range(self._workers_num):
+            task_r, task_w = mp.Pipe(duplex=False)
+            res_r, res_w = mp.Pipe(duplex=False)
+            sock_reader, sock_writer = socket.socketpair(family=socket.AF_UNIX)
```

Comment:
```suggestion
            sock_reader, sock_writer = multiprocessing.Pipe()
```

Reply:
done. But I wonder if creating an INET socket and than creating UNIX sockets based on it is sensible approach if we know we're running only on Unix.